src/Microsoft.Diagnostics.Monitoring.EventPipe/EventMonitoringPassthroughStream.cs

```diff
@@ -0,0 +1,260 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.EventPipe
+{
+    /// <summary>
+    /// A stream that can monitor an event stream which is compatible with <see cref="EventPipeEventSource"/>
+    /// for a specific event while also passing along the event data to a destination stream.
+    /// </summary>
+    internal sealed class EventMonitoringPassthroughStream : Stream
+    {
+        private readonly Action<TraceEvent> _onPayloadFilterMismatch;
+        private readonly Action<TraceEvent> _onEvent;
+        private readonly bool _callOnEventOnlyOnce;
+
+        private readonly Stream _sourceStream;
+        private readonly Stream _destinationStream;
+        private EventPipeEventSource _eventSource;
+
+        private readonly string _providerName;
+        private readonly string _eventName;
+
+        // The original payload filter of fieldName->fieldValue specified by the user. It will only be used to hydrate _payloadFilterIndexCache.
+        private readonly IDictionary<string, string> _payloadFilter;
+
+        // This tracks the exact indices into the provided event's payload to check for the expected values instead
+        // of repeatedly searching the payload for the field names in _payloadFilter.
+        private Dictionary<int, string> _payloadFilterIndexCache;
+
+
+        /// <summary>
+        /// A stream that can monitor an event stream which is compatible with <see cref="EventPipeEventSource"/>
+        /// for a specific event while also passing along the event data to a destination stream.
```

Comment:
This type feels like it couples solving two different problems and we should refactor it to split them. Does it complicate things much to have two classes?

1. PassthroughStream - this is a Stream that forwards whatever is read from it to a 2nd stream. It would include all the methods between lines 200-260 and you could create one like this:
```C#
PassthroughStream s = new PassthroughStream(sourceStream, destinationStream, bufferSize, leaveDestinationStreamOpen);
```

2. EventMonitor - this is an object which listens to a Stream of EventPipe data and invokes a callback when it spots a matching event.
```C#
EventMonitor e = new EventMonitor(providerName, eventName, filter, onEvent, onMismatch, inputStream, callOnlyOnce);
```

To use them together you would pass the Passthrough stream object as the inputStream parameter for EventMonitor.